File: reference/src/lossfunction/domain/portfolio.py
"""Portfolio aggregate — position aggregation over fills.

Exact arithmetic via Decimal; average price blends on buys and survives sells.
Realized P&L accrues on sells against the average price at sell time.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict

from lossfunction.domain.errors import InsufficientPositionError
from lossfunction.domain.types import OrderSide, Price, Quantity, Symbol


class Fill(BaseModel):
    """An execution that has happened (broker-reported)."""

    model_config = ConfigDict(frozen=True)

    client_order_id: str
    symbol: Symbol
    side: OrderSide
    quantity: Quantity
    price: Price


class PositionState(BaseModel):
    """Mutable-by-replacement position snapshot inside the portfolio."""

    model_config = ConfigDict(frozen=True)

    symbol: Symbol
    quantity: Quantity
    average_price: Price
    realized_pnl: Price = Price(0)


class Portfolio(BaseModel):
    """Aggregate of positions; updated only through `apply_fill`."""

    model_config = ConfigDict(frozen=True)

    positions: dict[str, PositionState] = {}

    def position(self, symbol: str) -> PositionState | None:
        return self.positions.get(symbol)

    @property
    def total_realized_pnl(self) -> Price:
        return Price(sum((p.realized_pnl for p in self.positions.values()), start=Price(0)))

    def apply_fill(self, fill: Fill) -> Portfolio:
        """Return a new Portfolio with the fill applied."""
        current = self.positions.get(fill.symbol)
        if fill.side is OrderSide.BUY:
            if current is None:
                new_position = PositionState(
                    symbol=fill.symbol,
                    quantity=fill.quantity,
                    average_price=fill.price,
                )
            else:
                total_cost = current.average_price * current.quantity + fill.price * fill.quantity
                new_quantity = current.quantity + fill.quantity
                new_position = PositionState(
                    symbol=fill.symbol,
                    quantity=new_quantity,
                    average_price=Price(total_cost / new_quantity),
                    realized_pnl=current.realized_pnl,
                )
            return self._with(new_position)

        if current is None or current.quantity < fill.quantity:
            held = 0 if current is None else current.quantity
            msg = f"insufficient position for {fill.symbol}: sell {fill.quantity}, held {held}"
            raise InsufficientPositionError(msg)

        realized = (fill.price - current.average_price) * fill.quantity
        remaining = current.quantity - fill.quantity
        # Fully-closed positions stay in the map with quantity 0 (mirrors KIS
        # inquire-balance, which shows same-day full sells as 0-quantity rows
        # until D-2); realized P&L is preserved on the entry.
        new_position = PositionState(
            symbol=fill.symbol,
            quantity=remaining,
            average_price=current.average_price,
            realized_pnl=current.realized_pnl + realized,
        )
        return self._with(new_position)

    def _with(self, position: PositionState) -> Portfolio:
        updated = self.positions.copy()
        updated[position.symbol] = position
        return Portfolio(positions=updated)

File: rust/lossfunction/src/domain/portfolio.rs
//! Portfolio aggregate — position aggregation over fills.
//!
//! Exact arithmetic via Decimal; average price blends on buys and survives
//! sells. Realized P&L accrues on sells against the average price at sell
//! time. Fully-closed positions stay with quantity 0 (KIS inquire-balance
//! parity; realized P&L preserved on the entry).

use std::collections::BTreeMap;

use rust_decimal::Decimal;

use crate::types::{OrderSide, Price, Quantity, Symbol};

use super::{blended_average, DomainError};

#[derive(Debug, Clone, PartialEq)]
pub struct Fill {
    pub client_order_id: String,
    pub symbol: Symbol,
    pub side: OrderSide,
    pub quantity: Quantity,
    pub price: Price,
}

#[derive(Debug, Clone, PartialEq)]
pub struct PositionState {
    pub symbol: Symbol,
    pub quantity: Quantity,
    pub average_price: Price,
    pub realized_pnl: Price,
}

/// Aggregate of positions; updated only through [`Portfolio::apply_fill`].
#[derive(Debug, Clone, Default, PartialEq)]
pub struct Portfolio {
    positions: BTreeMap<Symbol, PositionState>,
}

impl Portfolio {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn position(&self, symbol: &Symbol) -> Option<&PositionState> {
        self.positions.get(symbol)
    }

    pub fn positions(&self) -> impl Iterator<Item = &PositionState> {
        self.positions.values()
    }

    pub fn total_realized_pnl(&self) -> Price {
        self.positions
            .values()
            .map(|position| position.realized_pnl)
            .sum()
    }

    /// Return a new portfolio with the fill applied.
    pub fn apply_fill(&self, fill: Fill) -> Result<Portfolio, DomainError> {
        let mut next = self.clone();
        let entry = next.positions.get(&fill.symbol).cloned();
        match fill.side {
            OrderSide::Buy => {
                let updated = match entry {
                    None => PositionState {
                        symbol: fill.symbol.clone(),
                        quantity: fill.quantity,
                        average_price: fill.price,
                        realized_pnl: Decimal::ZERO,
                    },
                    Some(current) => PositionState {
                        symbol: fill.symbol.clone(),
                        quantity: current.quantity + fill.quantity,
                        average_price: blended_average(
                            current.quantity,
                            current.average_price,
                            fill.quantity,
                            fill.price,
                        ),
                        realized_pnl: current.realized_pnl,
                    },
                };
                next.positions.insert(fill.symbol.clone(), updated);
            }
            OrderSide::Sell => {
                let current = entry.ok_or_else(|| DomainError::InsufficientPosition {
                    symbol: fill.symbol.to_string(),
                    requested: fill.quantity,
                    held: 0,
                })?;
                if current.quantity < fill.quantity {
                    return Err(DomainError::InsufficientPosition {
                        symbol: fill.symbol.to_string(),
                        requested: fill.quantity,
                        held: current.quantity,
                    });
                }
                let realized = (fill.price - current.average_price) * Decimal::from(fill.quantity);
                let updated = PositionState {
                    symbol: fill.symbol.clone(),
                    quantity: current.quantity - fill.quantity,
                    average_price: current.average_price,
                    realized_pnl: current.realized_pnl + realized,
                };
                next.positions.insert(fill.symbol.clone(), updated);
            }
        }
        Ok(next)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::Symbol;

    fn symbol(code: &str) -> Symbol {
        Symbol::parse(code).unwrap()
    }

    fn fill(code: &str, side: OrderSide, qty: i64, price: i64) -> Fill {
        Fill {
            client_order_id: "c-1".into(),
            symbol: symbol(code),
            side,
            quantity: qty,
            price: Decimal::from(price),
        }
    }

    #[test]
    fn buy_creates_position_at_fill_price() {
        let portfolio = Portfolio::new()
            .apply_fill(fill("005930", OrderSide::Buy, 10, 80_000))
            .unwrap();
        let position = portfolio.position(&symbol("005930")).unwrap();
        assert_eq!(position.quantity, 10);
        assert_eq!(position.average_price, Decimal::from(80_000));
    }

    #[test]
    fn average_price_blends_across_buys() {
        let portfolio = Portfolio::new()
            .apply_fill(fill("005930", OrderSide::Buy, 10, 80_000))
            .unwrap()
            .apply_fill(fill("005930", OrderSide::Buy, 10, 82_000))
            .unwrap();
        let position = portfolio.position(&symbol("005930")).unwrap();
        assert_eq!(position.quantity, 20);
        assert_eq!(position.average_price, Decimal::from(81_000));
    }

    #[test]
    fn sell_keeps_average_price_and_books_realized_pnl() {
        let portfolio = Portfolio::new()
            .apply_fill(fill("005930", OrderSide::Buy, 10, 80_000))
            .unwrap()
            .apply_fill(fill("005930", OrderSide::Sell, 4, 85_000))
            .unwrap();
        let position = portfolio.position(&symbol("005930")).unwrap();
        assert_eq!(position.quantity, 6);
        assert_eq!(position.average_price, Decimal::from(80_000));
        assert_eq!(position.realized_pnl, Decimal::from(20_000)); // (85000-80000) * 4
        assert_eq!(portfolio.total_realized_pnl(), Decimal::from(20_000));
    }

    #[test]
    fn full_sell_leaves_zero_quantity_entry() {
        let portfolio = Portfolio::new()
            .apply_fill(fill("005930", OrderSide::Buy, 10, 80_000))
            .unwrap()
            .apply_fill(fill("005930", OrderSide::Sell, 10, 90_000))
            .unwrap();
        let position = portfolio.position(&symbol("005930")).unwrap();
        assert_eq!(position.quantity, 0);
        assert_eq!(position.realized_pnl, Decimal::from(100_000));
    }

    #[test]
    fn oversell_and_bare_sell_rejected() {
        let portfolio = Portfolio::new()
            .apply_fill(fill("005930", OrderSide::Buy, 5, 80_000))
            .unwrap();
        assert!(matches!(
            portfolio.apply_fill(fill("005930", OrderSide::Sell, 6, 85_000)),
            Err(DomainError::InsufficientPosition {
                requested: 6,
                held: 5,
                ..
            })
        ));
        assert!(matches!(
            Portfolio::new().apply_fill(fill("005930", OrderSide::Sell, 1, 80_000)),
            Err(DomainError::InsufficientPosition {
                requested: 1,
                held: 0,
                ..
            })
        ));
    }

    #[test]
    fn positions_are_independent() {
        let portfolio = Portfolio::new()
            .apply_fill(fill("005930", OrderSide::Buy, 10, 80_000))
            .unwrap()
            .apply_fill(fill("035420", OrderSide::Buy, 3, 41_000))
            .unwrap();
        assert_eq!(
            portfolio.position(&symbol("005930")).unwrap().average_price,
            Decimal::from(80_000)
        );
        assert_eq!(
            portfolio.position(&symbol("035420")).unwrap().average_price,
            Decimal::from(41_000)
        );
    }

    #[test]
    fn blending_handles_fractional_averages_exactly() {
        let portfolio = Portfolio::new()
            .apply_fill(fill("005930", OrderSide::Buy, 3, 10))
            .unwrap()
            .apply_fill(fill("005930", OrderSide::Buy, 4, 20))
            .unwrap();
        // (3*10 + 4*20) / 7 = 110/7 — exact Decimal, no float drift.
        assert_eq!(
            portfolio.position(&symbol("005930")).unwrap().average_price,
            Decimal::from(110) / Decimal::from(7)
        );
    }
}
